server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java

```diff
@@ -0,0 +1,168 @@
+import io.crate.metadata.FunctionName;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.metadata.pgcatalog.PgCatalogSchemaInfo;
+import io.crate.metadata.tablefunctions.TableFunctionImplementation;
+import io.crate.types.DataTypes;
+import io.crate.types.RowType;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.IntStream;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.DataTypes.INTEGER;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+public final class GenerateSubscripts<T> extends TableFunctionImplementation<T> {
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "generate_subscripts");
+    private static final RowType RETURN_TYPE = new RowType(List.of(INTEGER), List.of("subscripts"));
+    private static final Iterator<Row> NO_RESULTS_ITERATOR = Collections.emptyIterator();
+    private static final Comparator<Integer> KEEP_ORDER = (o1, o2) -> 0;
+
+
+    public static void register(TableFunctionModule module) {
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.BOOLEAN.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private GenerateSubscripts(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    private static List<?> extractNestedArray(List<?> array, int targetLevel) {
+        List<?> nested = array;
+        for (int level = 2; level <= targetLevel; level++) {
+            if (nested == null || nested.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || e instanceof List)) {
+                throw new IllegalArgumentException("expected a nested array at level " + level);
+            }
+            List<?> head = (List<?>) nested.get(0);
+            if (head == null || head.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || ((List<?>) e).size() == head.size())) {
```

Comment:
The same comment about using the stream API. It is probably not that necessary to create a new stream object to iterate over a list and do some predicate check. You can achieve that with the for-loop in a not less concise manner as well.